chore: Use InfoBadge for TabBarItem badge support

morning4coffe-dev opened the pull request:
GitHub Issue (If applicable): closes #1134

This PR adds support for using WinUI InfoBadge for TabBarItem while maintaining the support for old Badge properties. On Skia, the InfoBadge might be a little offset: https://github.com/unoplatform/uno/issues/10326

<!-- Link to relevant GitHub issue if applicable. All PRs should be associated with an issue (GitHub issue or internal) -->

## PR Type

What kind of change does this PR introduce?

- Bugfix
- Feature

## PR Checklist

Please check if your PR fulfills the following requirements:
- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Tested the changes where applicable:
	- [x] UWP
	- [x] WinUI
	- [ ] iOS
	- [x] Android
	- [x] WASM
	- [ ] MacOS
- [ ] Updated the documentation as needed:
	- [ ] [General Doc Update](https://github.com/unoplatform/uno.toolkit.ui/tree/main/doc)
	- [x] [Controls Doc Update](https://github.com/unoplatform/uno.toolkit.ui/tree/main/doc/controls)
	- [ ] [Extensions Doc Update](https://github.com/unoplatform/uno.toolkit.ui/tree/main/doc/helpers)
	- [ ] [controls-styles.md](https://github.com/unoplatform/uno.toolkit.ui/blob/main/doc/controls-styles.md)
	- [ ] [lightweight-styling.md (LightWeight Styling Resource Keys)](https://github.com/unoplatform/uno.toolkit.ui/blob/main/doc/lightweight-styling.md)
- [ ] [Runtime Tests and/or UI Tests](https://platform.uno/docs/articles/contributing/guidelines/creating-tests.html) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal)
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
     Please note that breaking changes are likely to be rejected -->

morning4coffe-dev committed:
chore: Use InfoBadge for TabBarItem badge support

agneszitte commented on src/Uno.Toolkit.UI/Controls/TabBar/TabBarItem.Properties.cs:
```diff
@@ -51,6 +51,17 @@ public Visibility BadgeVisibility
+		}
+
+		public static readonly DependencyProperty InfoBadgeProperty =
+			DependencyProperty.Register(nameof(InfoBadge), typeof(Control), typeof(TabBarItem), new PropertyMetadata(null, OnPropertyChanged));
```
@kazo0 what do you think?
If we do that @MartinZikmund then we should cover [all the properties for InfoBage](https://learn.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.infobadge?view=winui-2.8).

MartinZikmund replied:
You are right, maybe it is unnecessary complexity and one property is better than multiple

kazo0 replied:
It makes sense to just have an `InfoBadge` property and the you can customize that instance however you want

agneszitte replied:
I think this discussion can be resolved then @MartinZikmund, @morning4coffe-dev

morning4coffe-dev replied:
Agreed 👍

kazo0 commented on src/Uno.Toolkit.UI/Controls/TabBar/TabBarItem.Properties.cs:
```diff
@@ -51,6 +51,17 @@ public Visibility BadgeVisibility
 			DependencyProperty.Register("BadgeValue", typeof(string), typeof(TabBarItem), new PropertyMetadata(default(string?), OnPropertyChanged));
 		#endregion
 
+		#region InfoBadge
```
Are we planning on removing the BadgeValue and BadgeVisibility properties and introduce this as a breaking change? I feel like this might be confusing to have InfoBadge alongside the existing Badge related properties and have them work independently. 

I'd say we either get rid of them or forward their functionality to the InfoBadge property

kazo0 replied:
fyi @MartinZikmund / @agneszitte

kazo0 replied:
My vote is to remove them and document a breaking change

agneszitte replied:
I think the same @kazo0 on my side and in addition to documenting the breaking change, this commit should be properly updated to mention a breaking change as well with conventional commits.
Do you think the same  @MartinZikmund, @Xiaoy312 ?

kazo0 replied:
Technically they can be data-bound but usually not. It suffers from [this issue](https://github.com/unoplatform/uno.toolkit.ui/issues/1129) if you were to try to use the TabBarItem from a DataTemplate

morning4coffe-dev replied:
Decided to not make a breaking change with the `InfoBadge` change at the moment according to conversation we had on standup. Now, users can use both and we can deprecate `BadgeVisibility` and `BadgeValue` in later releases.

morning4coffe-dev committed:
chore: Implement BadgeVisibility and BadgeValue with WinUI InfoBadge

Decided to not make a breaking change with the InfoBadge change at the moment. Now, users can use both and we can deprecate BadgeVisibility and BadgeValue in later releases.

morning4coffe-dev committed:
chore: Adjust placement of InfoBadge

agneszitte commented on src/Uno.Toolkit.UI/Controls/TabBar/TabBarItem.Properties.cs:
```diff
@@ -29,26 +27,61 @@ partial class TabBarItem
+			DependencyProperty.Register(nameof(BadgeValue), typeof(string), typeof(TabBarItem), new PropertyMetadata(default(string?), OnPropertyChanged));
+		#endregion
+
+		#region InfoBadge
```
@morning4coffe-dev the related documentation should be updated as well, please https://github.com/unoplatform/uno.toolkit.ui/tree/main/doc

morning4coffe-dev replied:
Added the `InfoBadge` there, thank you @agneszitte. Should I also remove the old properties from the table or mark them obsolete somehow?

agneszitte replied:
@kazo0, @Xiaoy312 based on this previous comment: https://github.com/unoplatform/uno.toolkit.ui/pull/1142#discussion_r1623935490
![image](https://github.com/unoplatform/uno.toolkit.ui/assets/16295702/7fb78067-af37-4b8b-aa9f-567b157fa777)
Since we are not doing a breaking change, I don't think we should remove them from the doc, but only mention them as deprecated right?

morning4coffe-dev committed:
docs: Add `InfoBadge` option to docs

agneszitte commented on src/Uno.Toolkit.UI/Controls/TabBar/TabBarItem.Properties.cs:
```diff
@@ -29,26 +27,61 @@ partial class TabBarItem
 		#endregion
 
+		// UNO TODO: Deprecate and remove BadgeVisibility and BadgeValue properties and use InfoBadge instead
 		#region BadgeVisibility
```
@kazo0, @Xiaoy312, @MartinZikmund 
Can we have something like [Obsolete("This property is deprecated. Use `InfoBadge` instead.", true)] for `BadgeVisibility` and `BadgeValue` here ?

agneszitte replied:
Related issue for reference: https://github.com/unoplatform/uno.csharpmarkup/issues/741
@morning4coffe-dev like we talked about, you add the [Obsolete] like we should have but comment it with the link of the related uno issue.

morning4coffe-dev committed:
chore: Note deprecated properties

morning4coffe-dev committed:
chore: Add Deprecated attribute

morning4coffe-dev committed:
chore: Add missing System using

agneszitte commented on doc/controls/TabBarAndTabBarItem.md:
```diff
@@ -47,14 +47,16 @@ xmlns:utu="using:Uno.Toolkit.UI"
+| `InfoBadge`        | `Control`    | Gets or sets the `InfoBadge` control placed on the top right corner of the `TabBarItem`.                                                                                                                                                                                                                          |
 | `IsSelectable`     | `bool`        | Gets or sets whether the `TabBarItem` can be selected.                                                                                                                                                                                                                                         |
 
+> [!NOTE]
+> The properties `BadgeVisibility` and `BadgeValue` have been deprecated. They were previously used to control the badge visibility and value on the `TabBarItem`. If you need similar functionality, consider using the `InfoBadge` property. The deprecated properties were only supported by the Material Theme Toolkit Library with `BottomTabBarItemStyle` and `VerticalTabBarItemStyle`.
```
@kazo0, @Xiaoy312, @MartinZikmund  what do think?
I though maybe we could keep the table as is but with an `(Obsolete)` after the name of the property in the first column in addition to this note. So we don't lose the type and description info as needed (since we are not doing a breaking change here).

MartinZikmund replied:
I'm ok with this description, but your improvement is useful, let's add the Obsolete mention, or an asterisk for example and then start this text with asterisk to show it explains what the asterisk means

morning4coffe-dev committed:
Merge branch 'main' into dev/doti/add-infobadge-tabbaritem

morning4coffe-dev committed:
chore: Comment out Obsolete

morning4coffe-dev committed:
chore: Minor InfoBadge adjustments

morning4coffe-dev committed:
docs: Adjust TabBarItem badge

morning4coffe-dev committed:
Merge branch 'main' into dev/doti/add-infobadge-tabbaritem

morning4coffe-dev commented:
@MartinZikmund @agneszitte @kazo0 Can we merge this now? The `InfoBadge` alignment issues on are now fixed with Skia and WASM is off only very slightly and it's not much noticable.

morning4coffe-dev commented on src/Uno.Toolkit.UI/Controls/TabBar/TabBarItem.Properties.cs:
```diff
@@ -29,26 +28,68 @@ partial class TabBarItem
 			DependencyProperty.Register(nameof(Icon), typeof(IconElement), typeof(TabBarItem), new PropertyMetadata(null, OnPropertyChanged));
 		#endregion
 
+		// UNO TODO: Deprecate and remove BadgeVisibility and BadgeValue properties and use InfoBadge instead
```
Thanks @kazo0, I will take a look next week 👍

morning4coffe-dev replied:
Tracked in #1626. The TabBar documentation now recommends InfoBadge for new code and records the major-version migration.

Commit:
chore: Merge latest main

Commit:
fix(tabbar): Preserve legacy badge behavior

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/unoplatform/uno.toolkit.ui?pullRequest=1142) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/unoplatform/uno.toolkit.ui?pullRequest=1142) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: morning4coffe-dev<br/>:x: Dominik Titl<br/><hr/>**Dominik Titl** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/unoplatform/uno.toolkit.ui?pullRequest=1142) it.</sub>

morning4coffe-dev commented on src/library/Uno.Toolkit.Material/Styles/Controls/v2/InfoBadge.xaml:
```diff
@@ -0,0 +1,72 @@
+	<!-- Large InfoBadge-->
+	<x:Double x:Key="MaterialLargeInfoBadgeBottomTabBarItemHeight">16</x:Double>
+	<Thickness x:Key="MaterialLargeInfoBadgeBottomTabBarItemMargin">32,4,0,0</Thickness>
+	<CornerRadius x:Key="MaterialLargeInfoBadgeBottomTabBarItemCornerRadius">8</CornerRadius>
```
Moved the InfoBadge lightweight styling resources into the Default theme dictionary.

kazo0 commented:
@morning4coffe-dev can we also add the support for it in the Simple Theme styles for TabBarItem? :)
